### PR TITLE
Adding decoder, please check

### PR DESCRIPTION
**Pull request by ardavan-afshar**
Does Tim Dettmers approve?

### CODE REVIEW

**Review by mfdowland — APPROVED** [review]
yh it's ok i guess

File: decoder.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+            self.encoder = Encoder(vocab_size, embedding_dim, ff_dim, num_heads, num_layers, dropout)
+        self.decoder = Decoder(vocab_size, embedding_dim, ff_dim, num_heads, num_layers, dropout, with_encoder)
+
+    def forward(self, text):
+        if self.with_encoder == True:
+            encoder_output = self.encoder(text)
+            probabilities = self.decoder(text, encoder_output)
+        else:    
+            probabilities = self.decoder(text, encoder_output = None)
+        return x
```
Comment (mantagen) [inline decoder.py]: what is x?

**Review by mantagen — CHANGES_REQUESTED** [review]

File: decoder.py [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+    def forward(self, text):
+        embedded_text = self.embed(text)
+        positionally_embedded_text = embedded_text + self.positional_embedding(text)
+        return positionally_embedded_text
+
+
+# In[79]:
+
+
+# emb = EmbeddingLayer(8, 5)
```
Comment (mantagen) [inline decoder.py]: could you delete these redundant comments?

**Review by mantagen — CHANGES_REQUESTED** [review]

File: decoder.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+# (Masked) Multi-head attention
+
+# In[ ]:
+
+
+class MultiHeadAttention(nn.Module):
+    def __init__(self, d_model, num_heads, forward_masked):
+        super(MultiHeadAttention, self).__init__()
+        self.d_model = d_model
```
Comment (mantagen) [inline decoder.py:52]: how about we are consistent across the codebase with naming convention for dimenality?

d_model? head_dim?

Comment (mantagen) [reply decoder.py:52]: ff_dim?

Comment (mantagen) [reply decoder.py:52]: embedding_dim?

**Review by louisc-s — APPROVED** [review]

**Review by mantagen — COMMENTED** [review]
Also, at the bottom of the file, under ` === "main"`, could do a simple inference with no weights loaded -- just as a form of sanity test

File: decoder.py (hunk 3 above)
Comment (ardavan-afshar) [reply decoder.py:52]: Fair enough, maybe later.

**Review by mantagen — APPROVED** [review]

**Pull request merged**